sql: ensure that hidden columns are effectively hidden

knz opened the pull request:
CockroachDB supports "hidden columns" in logical plans -- columns that
can be selected explicitly if their name is known, but that are
otherwise invisible: not visible to star expansion, and (in principle)
*not visible to clients* except in the output of EXPLAIN.

This was true for the longest time, but became broken at some point
during summer 2017 when we introduced result streaming: values for
hidden columns in the top-level of queries became erroneously streamed
alongside those for visible columns. This escaped testing because we
didn't have tests for this.

This patch corrects the issue by ensuring that a renderNode
(projection) is added at the top of the logical plan to only render
visible columns if there are any hidden columns underneath. The
solution uses an alteration of the logical plan (as opposed to
e.g. special code in the executor/pgwire) so that the hidden columns
can be properly optimized away from the logical plans if they are not
rendered in the result.

Release note (bug fix): the "JSON" column in the output of
EXPLAIN(DISTSQL) is now properly hidden by default. It can be shown
using `SELECT *, JSON FROM [EXPLAIN(DISTSQL) ...]`.

Fixes  #21089.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21154)
<!-- Reviewable:end -->

knz committed:
sql: ensure that hidden columns are effectively hidden

CockroachDB supports "hidden columns" in logical plans -- columns that
can be selected explicitly if their name is known, but that are
otherwise invisible: not visible to star expansion, and (in principle)
*not visible to clients* except in the output of EXPLAIN.

This was true for the longest time, but became broken at some point
during summer 2017 when we introduced result streaming: values for
hidden columns in the top-level of queries became erroneously streamed
alongside those for visible columns. This escaped testing because we
didn't have tests for this.

This patch corrects the issue by ensuring that a renderNode
(projection) is added at the top of the logical plan to only render
visible columns if there are any hidden columns underneath. The
solution uses an alteration of the logical plan (as opposed to
e.g. special code in the executor/pgwire) so that the hidden columns
can be properly optimized away from the logical plans if they are not
rendered in the result.

Release note (bug fix): the "JSON" column in the output of
EXPLAIN(DISTSQL) is now properly hidden by default. It can be shown
using `SELECT *, JSON FROM [EXPLAIN(DISTSQL) ...]`.

RaduBerinde commented:
LGTM, thanks for the fix!

knz commented:
TFYR!
